shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1965,13 +1966,26 @@ - (void)updateEditingStateWithDelta:(flutter::TextEditingDelta)delta {
     @"composingExtent" : @(composingExtent),
   };
 
-  NSDictionary* deltas = @{
-    @"deltas" : @[ deltaToFramework ],
-  };
+  [_pendingDeltas addObject:deltaToFramework];
```

Comment:
Now that there's batching, what happens if `oldText` is different in the same batch? For example when there's a framework change happened in between?

Reply:
Do you mean in this situation:

Batch: Insertion, Deletion, Insertion
After first insertion delta is applied, the framework changes the editing state in response to the insertion?

Reply:
If the engine applies an update from the framework, after the block is scheduled and before the block is invoked. In theory that can happen right?

Reply:
I see what you mean. In theory that can happen something like the following example.
Type A followed by B:

Engine types A.
Engine schedules delta block to run.
Framework types C for x reason.
Engine has C.
Engine types B.
Engine has CB and sends to framework, Expected ACB?

Deltas sent:
InsertionA, oldText: ''
InsertionB, oldText:'C'.

Here the user will not really be able to recover. I don't think this is too different from the IME-Framework out of sync issues we usually face. What do you think about creating some method that signals the framework that a batch has been scheduled, and that it probably is not safe to make any framework side changes during this time.

Reply:
In your example the batch is going to be partially applied by the framework right? The framework will apply changes in the batch until the point where `oldText` mismatches? Shouldn't it be all or nothing, when there's on inconsistent "oldText", when framework rejects the entire batch?

Reply:
I think the framework will ignore any difference in the `oldText`. https://github.com/flutter/flutter/blob/f468f3366c26a5092eb964a230ce7892fda8f2f8/packages/flutter/lib/src/services/text_editing_delta.dart#L324-L334 . A `TextEditingDelta` applies the delta to the `oldText` to follow a last write wins policy. A user can still prevent the application of a delta by checking the `oldText` with their current value themselves before calling `apply`. A user can make it `all or nothing`  but we don't do this by default.

Reply:
Sorry missed the reply. 

> A user can make it all or nothing but we don't do this by default

So the current implementation in the framework doesn't guarantee that a delta batch is atomically applied? That sounds like the same issue could still happen if a framework edit is sent to the engine while the engine is replacing a double-dash?

Reply:
Oh the framework edit is going to be completely ignored in that case?

Reply:
> So the current implementation in the framework doesn't guarantee that a delta batch is atomically applied? That sounds like the same issue could still happen if a framework edit is sent to the engine while the engine is replacing a double-dash?

Yes in theory a framework edit can still come in during the batch accumulation and interrupt the double-dash replacement. I don't think we can avoid this case but I don't think it is less likely. Before this change the double-dash replacement sequence would come in as 3 separate delta batches, so the user could potentially react to the 1st delta batch and make some edit and disrupt the sequence. After this change the sequence comes in as one delta batch, so the user can avoid making an edit until they fully apply the delta batch.

> Oh the framework edit is going to be completely ignored in that case?
No, it is still applied by the engine. In the example I stated above the edit that is ignored is the 1st engine edit that types 'A'.